### PR TITLE
Fix RetinaNet training on TPU.

### PR DESCRIPTION
**Pull request by tanzhenyu**
This gives mAP ~34.0

Fixes https://github.com/keras-team/keras-cv/issues/1075

### CODE REVIEW

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 1]
```diff
@@ -1,15 +1,37 @@
...
+Date created: 2022/09/27
+Last modified: 2022/12/08
+Description: Use KerasCV to train a RetinaNet on Pascal VOC 2007.
+"""
 import resource
+import sys
 
 import tensorflow as tf
 import tensorflow_datasets as tfds
+from absl import flags
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:26]: order wise, the absl should appear before tf.

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 2]
```diff
@@ -38,9 +60,38 @@
 ]
 class_mapping = dict(zip(range(len(class_ids)), class_ids))
 
+flags.DEFINE_string(
+    "weights_path",
+    "weights_{epoch:02d}.h5",
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:65]: It seems that this is not just a path, but a formatted string for the file name. Should we make it clear in the flag description?

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 3]
```diff
@@ -38,9 +60,38 @@
...
+    "Directory which will be used to store weight checkpoints.",
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:66]: h5 seems to be a file, not a directory.

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 4]
```diff
@@ -38,9 +60,38 @@
...
+)
+flags.DEFINE_string(
+    "tensorboard_path",
+    "logs",
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:70]: The default value here is bit weird, and should it have a relative location?

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 5]
```diff
@@ -38,9 +60,38 @@
...
+    "Directory which will be used to store tensorboard logs.",
+)
+FLAGS = flags.FLAGS
+FLAGS(sys.argv)
+
+# parameters from RetinaNet [paper](https://arxiv.org/abs/1708.02002)
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:76]: What's the "parameters" here?

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 6]
```diff
@@ -202,41 +253,85 @@ def apply(inputs):
...
+
+def pad_fn(image, boxes):
+    return image, boxes.to_tensor(default_value=-1.0)
+
+
 train_ds = train_ds.map(
     proc_train_fn("xywh", image_size), num_parallel_calls=tf.data.AUTOTUNE
 )
 train_ds = train_ds.apply(
     tf.data.experimental.dense_to_ragged_batch(BATCH_SIZE, drop_remainder=True)
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:317]: Shouldn't we use the global batch size here for the original dataset?

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 7]
```diff
@@ -202,41 +253,85 @@ def apply(inputs):
...
 )
+
+train_ds = train_ds.map(pad_fn, num_parallel_calls=tf.data.AUTOTUNE)
+train_ds = train_ds.shuffle(8)
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:321]: Should this value change based on the batch size?

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 8]
```diff
@@ -202,41 +253,85 @@ def apply(inputs):
...
+        raw_image = inputs["image"]
+        raw_image = tf.cast(raw_image, tf.float32)
+
+        img_size = tf.shape(raw_image)
+        height = img_size[0]
+        width = img_size[1]
+
+        target_height = tf.cond(
+            height > width,
+            lambda: 640.0,
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:272]: Probably lift this 640 out as a constant.

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 9]
```diff
@@ -202,41 +253,85 @@ def apply(inputs):
...
+            images=image,
+            source="xyxy",
+            target=bounding_box_format,
+        )
+        gt_classes = tf.cast(inputs["objects"]["label"], tf.float32)
+        gt_classes = tf.expand_dims(gt_classes, axis=-1)
+
+        bounding_boxes = tf.concat([gt_boxes, gt_classes], axis=-1)
+
+        return image, bounding_boxes
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:304]: Should we return dict (same as the input), or just the tuple here.

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 10]
```diff
@@ -250,190 +345,62 @@ def unpackage_dict(inputs):
...
+        # pixel range (0, 255) or if you have already rescaled your inputs to the range
+        # (0, 1).  In our case, we feed our model images with inputs in the range (0, 255).
+        include_rescaling=True,
+    )
+    # Fine-tuning a RetinaNet is as simple as setting backbone.trainable = False
+    model.backbone.trainable = False
+
+    metrics = [
+        keras_cv.metrics.COCOMeanAveragePrecision(
+            class_ids=range(21),
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:370]: maybe we should change 21 to len(class_ids)

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 10 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py]: Actually no longer needed. Removed.

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 9 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:305]: the input needs to be tuple. The boxes will be splitted into dict after RetinaNet supports that.

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 8 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py]: Done.

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 7 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py]: Good catch. Done

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 6 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py]: Good catch. Done

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 5 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:76]: Everything from batch size, learning rate, to image size, etc.

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 4 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:70]: By default it is using relative path

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 3 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:66]: Changing to `weights_name` (compare to file_name)

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 2 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:65]: Changing to `weights_name`, compare to file_name

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 1 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:26]: Done.

Comment (qlzh727) [reply examples/training/object_detection/pascal_voc/retina_net.py:26]: This doesn't seems to be updated.

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 11]
```diff
@@ -202,41 +254,85 @@ def apply(inputs):
...
+train_ds = train_ds.shuffle(8 * strategy.num_replicas_in_sync)
+train_ds = train_ds.prefetch(2)
+
 eval_ds = eval_ds.map(
-    proc_train_fn(bounding_box_format="xywh", img_size=image_size),
+    proc_eval_fn(bounding_box_format="xywh", target_size=image_size),
     num_parallel_calls=tf.data.AUTOTUNE,
 )
 eval_ds = eval_ds.apply(
     tf.data.experimental.dense_to_ragged_batch(BATCH_SIZE, drop_remainder=True)
```
Comment (qlzh727) [inline examples/training/object_detection/pascal_voc/retina_net.py:330]: same here, it should be the global batch size.

**Review by qlzh727 — APPROVED** [review]

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 1 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py:26]: So that seems to be `shell/format.sh` auto-correcting it

File: examples/training/object_detection/pascal_voc/retina_net.py (hunk 11 above)
Comment (tanzhenyu) [reply examples/training/object_detection/pascal_voc/retina_net.py]: Done.
